File: sdk/dotnet/State/RemoteStateReference.cs
// Copyright 2016-2020, Pulumi Corporation.

using System.Collections.Immutable;
using Pulumi.Serialization;

namespace Pulumi.Terraform.State
{
    /// <summary>
    /// Manages a reference to a Terraform Remote State. The root outputs of the remote state are available
    /// via the <see cref="Outputs"/> property or the <see cref="GetOutput"/> method.
    /// </summary>
    public class RemoteStateReference : CustomResource
    {
        /// <summary>
        /// The root outputs of the referenced Terraform state. 
        /// </summary>
        [Output("outputs")]
        public Output<ImmutableDictionary<string, object>> Outputs { get; private set; } = null!;

        /// <summary>
        /// Create a RemoteStateReference resource with the given unique name, arguments, and options.
        /// </summary>
        /// <param name="name">The unique name of the remote state reference.</param>
        /// <param name="args">The arguments to use to populate this resource's properties.</param>
        /// <param name="options">A bag of options that control this resource's behavior.</param>
        public RemoteStateReference(string name, RemoteStateReferenceArgs args, CustomResourceOptions? options = null)
            : base("terraform:state:RemoteStateReference", 
                name, 
                args, 
                CustomResourceOptions.Merge(options, new CustomResourceOptions { Id = name }))
        {
        }

        /// <summary>
        /// Fetches the value of a root output from the Terraform Remote State.
        /// </summary>
        /// <param name="name">The name of the output to fetch. The name is formatted exactly as per
        /// the "output" block in the Terraform configuration.</param>
        /// <returns></returns>
        public Output<object> GetOutput(Input<string> name)
            => Output.Tuple(name.ToOutput(), Outputs).Apply(v => v.Item2[v.Item1]);
    }
    
    /// <summary>
    /// The base type for arguments for constructing a RemoteStateReference resource.
    /// </summary>
    public abstract class RemoteStateReferenceArgs : ResourceArgs
    {
        /// <summary>
        /// A constant describing the name of the Terraform backend, used as the discriminant
        /// for the union of backend configurations.
        /// </summary>
        public abstract Input<string> BackendType { get; }
    }
}

File: sdk/go/state/remote_state_reference.go
package state

import (
	"errors"

	"github.com/pulumi/pulumi/sdk/v2/go/pulumi"
)

// RemoteStateReference is a resource which allows reading the output from local or remote Terraform state files
type RemoteStateReference struct {
	pulumi.CustomResourceState

	// Outputs is a map of the outputs from the Terraform state file
	Outputs pulumi.MapOutput `pulumi:"outputs"`
}

// RemoteStateReferenceArgs is an interface implemented by all valid backend type Arg structs
type RemoteStateReferenceArgs interface {
	toInternalArgs() pulumi.Input
	validateArgs() error
}

// NewRemoteStateReference manages a reference to a Terraform Remote State.
// The root outputs of the remote state are available via the `Outputs` property
func NewRemoteStateReference(ctx *pulumi.Context, name string, args RemoteStateReferenceArgs,
	opts ...pulumi.ResourceOption) (*RemoteStateReference, error) {

	if args == nil {
		return nil, errors.New("missing required argument 'args'")
	}

	err := args.validateArgs()
	if err != nil {
		return nil, err
	}

	internalArgs := args.toInternalArgs()

	var resource RemoteStateReference
	err = ctx.ReadResource("terraform:state:RemoteStateReference", name, pulumi.ID(name), internalArgs, &resource, opts...)
	if err != nil {
		return nil, err
	}
	return &resource, nil
}
